dask_kubernetes/operator/daskcluster.py

```diff
@@ -0,0 +1,274 @@
+    worker_pods = api.create_namespaced_custom_object(
+        group="kubernetes.dask.org",
+        version="v1",
+        plural="daskworkergroups",
+        namespace=namespace,
+        body=data,
+    )
+    logger.info(
+        f"A worker group has been created called {data['metadata']['name']} in {namespace} \
+        with the following config: {data['spec']}"
+    )
+
+
+@kopf.on.create("daskworkergroup")
+async def daskworkergroup_create(spec, name, namespace, logger, **kwargs):
+    api = kubernetes.client.CoreV1Api()
+
+    cluster = kubernetes.client.CustomObjectsApi().list_cluster_custom_object(
+        group="kubernetes.dask.org", version="v1", plural="daskclusters"
+    )
+    scheduler_name = cluster["items"][0]["metadata"]["name"]
+    num_workers = spec["replicas"]
+    for i in range(num_workers):
+        data = build_worker_pod_spec(
+            name, namespace, spec.get("image"), uuid4().hex, scheduler_name
+        )
+        kopf.adopt(data)
+        worker_pod = api.create_namespaced_pod(
+            namespace=namespace,
+            body=data,
+        )
+    logger.info(f"{num_workers} Worker pods in created in {namespace}")
+
+
+@kopf.on.update("daskworkergroup")
+async def daskworkergroup_update(spec, name, namespace, logger, **kwargs):
+    api = kubernetes.client.CoreV1Api()
+
+    scheduler = kubernetes.client.CustomObjectsApi().list_cluster_custom_object(
+        group="kubernetes.dask.org", version="v1", plural="daskclusters"
+    )
+    scheduler_name = scheduler["items"][0]["metadata"]["name"]
+    workers = api.list_namespaced_pod(
+        namespace=namespace,
+        label_selector=f"dask.org/workergroup-name={name}",
+    )
+    current_workers = len(workers.items)
+    desired_workers = spec["replicas"]
+    workers_needed = desired_workers - current_workers
+    if workers_needed > 0:
+        for i in range(workers_needed):
+            data = build_worker_pod_spec(
+                name, namespace, spec.get("image"), uuid4().hex, scheduler_name
+            )
+            kopf.adopt(data)
+            worker_pod = api.create_namespaced_pod(
+                namespace=namespace,
+                body=data,
+            )
+        logger.info(f"Scaled worker group {name} up to {spec['replicas']} workers.")
+    if workers_needed < 0:
+        # TODO: Replace address localhost with the scheduler service name
+        scheduler = rpc("localhost:8786")
+        worker_ids = await scheduler.workers_to_close(
+            n=-workers_needed, attribute="name"
+        )
+        logger.info(f"Workers to close: {worker_ids}")
+        for wid in worker_ids:
+            worker_pod = api.delete_namespaced_pod(
+                name=wid,
+                namespace=namespace,
+            )
+        logger.info(f"Scaled worker group {name} down to {spec['replicas']} workers.")
+
+
+@kopf.on.delete("daskcluster")
+async def daskcluster_delete(spec, name, namespace, logger, **kwargs):
```

Comment:
Instead of handling the delete manually you could rely on [Kubernetes Owner References](https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/) for the cleanup. 
Basically you could set the `daskcluster` resource as an owner for everything else you create here (the pods, services and daskworkergroups) and when it gets deleted kubernetes will take care of the rest.

I think you are already doing this for the native kubernetes objects you create in here (via the `kopf.adopt` calls).

Reply:
Yeah we are aware of this. There was a discussion about it in [#406](https://github.com/dask/dask-kubernetes/pull/406#pullrequestreview-885663075). We were having a little trouble getting the adoption to work when scaling up and it was blocking the PR. But we decided to just merge in this for now and figure out the adoption issues later.